### PR TITLE
DAOS-14529 control: Fix config usage in daos_server cmds

### PR DESCRIPTION
**Pull request by tanabarr**
Centralize processing of --ignore-config flag in daos_server
stand-alone commands by moving logic into embedded type cfgCmd
LoadConfig() methods. Additionally show log buffer on case failure for
JSON unit tests and extend unit testing to properly cover -o and
--ignore-config opts.

Make -o option subcommand specific by moving ConfigPath field from the
mainOpts struct to cfgCmd so that only commands that embed the latter
will inherit the -o flag. Specifically daos_server config generate
command will no longer support -o with the result that config file
input and output behaviour will be unambiguous.

One notable consequence of the change is that -o|--config needs to be
specified after the subcommand and not before.

Fixes: DAOS-12359 DAOS-12891
Features: control
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by kjacque — APPROVED** [review]
Nice cleanup, LGTM

**Review by mjmac — APPROVED** [review]

**Review by knard-intel — DISMISSED** [review]

**Review by kjacque — DISMISSED** [review]
Nice -- I appreciate the consolidation of the config related logic into one place.

**Review by kjacque — DISMISSED** [review]

**Comment by tanabarr** [comment]
error in https://github.com/daos-stack/daos/actions/runs/8800672314/job/24152350268?pr=14067 is because --config now needs to be specified after the "start" subcommand (daos_server start --config ... as opposed to daos_server --config start in nlt script)

**Review by kjacque — APPROVED** [review]
Hmm, you should probably send an email to the mailing list about the need to change the order of parameters. This may take some people by surprise.

**Review by knard-intel — APPROVED** [review]
LGTM

File: src/control/server/config/server.go [hunk 1]
```diff
@@ -1,5 +1,5 @@
 //
-// (C) Copyright 2020-2023 Intel Corporation.
+// (C) Copyright 2020-2024 Intel Corporation.
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/server/config/server.go:2]: missed this on the last push, will address on the next

**Review by kjacque — DISMISSED** [review]

File: src/tests/ftest/util/server_utils_base.py [hunk 2]
```diff
@@ -86,6 +79,11 @@ def get_sub_command_class(self):
             # Use the sub_command parameter from the test's yaml
             sub_command = self.sub_command.value
 
+        # If specified use the configuration file from the YamlParameters object as the default
+        default_config = None
```
Comment (knard-intel) [inline src/tests/ftest/util/server_utils_base.py:83]: I am probably missing something, but it seems that it default_config sould be initialized in the following way.
```suggestion
        default_config = DEFAULT_CONFIG_FILE 
```

Comment (tanabarr) [reply src/tests/ftest/util/server_utils_base.py:83]: I think in the context of the test the --config option should only be specified if the caller provides it. @phender to confirm

**Review by daltonbohning — COMMENTED** [review]
@phender 

Check failure on line 470 in src/tests/ftest/util/server_utils.py
GitHub Actions / Pylint check
no-member, Instance of 'DaosServerCommand' has no 'config' member


From
```
        cmd = DaosServerCommand(self.manager.job.command_path)
        cmd.run_user = "daos_server"
        cmd.debug.value = False
        cmd.config.value = get_default_config_file("server")
        self.log.info("Support collect-log on servers: %s", str(cmd))
        cmd.set_command(("support", "collect-log"), **kwargs)
```

File: src/tests/ftest/util/server_utils_base.py (hunk 2 above)
Comment (daltonbohning) [reply src/tests/ftest/util/server_utils_base.py:83]: The code path ultimately calls `get_default_config_file`, and it looks like `DEFAULT_CONFIG_FILE` isn't used anymore. Probably old code that should be removed (pre-existing)

Comment (knard-intel) [reply src/tests/ftest/util/server_utils_base.py:83]: @daltonbohning thanks for the explanation.

**Review by knard-intel — DISMISSED** [review]

**Comment by tanabarr** [comment]
@phender support collect log presumably uses a helper that still needs updating https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-14067/18/tests/ I will investigate

File: src/tests/ftest/util/server_utils.py [hunk 3]
```diff
@@ -467,9 +467,9 @@ def support_collect_log(self, **kwargs):
         cmd = DaosServerCommand(self.manager.job.command_path)
         cmd.run_user = "daos_server"
         cmd.debug.value = False
-        cmd.config.value = get_default_config_file("server")
```
Comment (daltonbohning) [inline src/tests/ftest/util/server_utils.py:470]: I think this should work - we could also do
```
kwargs['config'] = get_default_config_file("server")
And let the `set_command` below continue normally
```

Comment (tanabarr) [reply src/tests/ftest/util/server_utils.py:470]: Isn't that the same thing? `f(a=foo, **kwargs)` is essentially the same as `kwargs['a']=foo; f(**kwargs)` I would have thought

Comment (daltonbohning) [reply src/tests/ftest/util/server_utils.py:470]: Yes, the same functionally :) No need to change

File: src/tests/ftest/util/server_utils.py [hunk 4]
```diff
@@ -467,9 +467,9 @@ def support_collect_log(self, **kwargs):
...
+        cmd.set_command(("support", "collect-log"),
+                        "config"=get_default_config_file("server"), **kwargs)
```
Comment (daltonbohning) [inline src/tests/ftest/util/server_utils.py:471]: As an argument, no quotes
```suggestion
                        config=get_default_config_file("server"), **kwargs)
```

Comment (phender) [inline src/tests/ftest/util/server_utils.py]: As we discussed:
```
        kwargs['config'] = get_default_config_file("server")
        cmd.set_command(("support", "collect-log"), **kwargs)
```

**Review by phender — CHANGES_REQUESTED** [review]

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

File: src/tests/ftest/util/server_utils.py (hunk 4 above)
Comment (tanabarr) [reply src/tests/ftest/util/server_utils.py]: as Phil points out, there is a subtle difference in the case where the config value is set in kwargs and supplied in set_command parameters

**Review by phender — CHANGES_REQUESTED** [review]

File: src/tests/ftest/util/server_utils.py [hunk 5]
```diff
@@ -467,9 +467,9 @@ def support_collect_log(self, **kwargs):
         cmd = DaosServerCommand(self.manager.job.command_path)
         cmd.run_user = "daos_server"
         cmd.debug.value = False
-        cmd.config.value = get_default_config_file("server")
-        self.log.info("Support collect-log on servers: %s", str(cmd))
+        kwargs['config'] = get_default_config_file("server")
         cmd.set_command(("support", "collect-log"), **kwargs)
+        self.log.info("Support collect-log on servers: %s", str(cmd))
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/tests/ftest/util/server_utils.py:472]: it was just moved from the previous line, I didn't add it

**Review by phender — DISMISSED** [review]
As Tom pointed out to me what I thought was an added line already exitsted.  Ftest changes look good.

**Review by kjacque — DISMISSED** [review]

**Review by knard-intel — DISMISSED** [review]

**Comment by daltonbohning** [comment]
@phender It looks like the config needs to be passed down another layer? E.g.:
https://github.com/daos-stack/daos/blob/0630b4d1fbcd5a0caaa7b2cd4d57f89ecd2ba3b2/src/tests/ftest/util/server_utils_base.py#L557-L563

**Comment by phender** [comment]
@tanabarr I've pushed changes to the DaosServerCommand class that should now properly place the `--config` argument.  I had them in the incorrect position previously.

**Review by tanabarr — COMMENTED** [review]
thanks for the update

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

**Review by kjacque — APPROVED** [review]

**Pull request merged**